torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -207,7 +217,27 @@ struct formatter<::addrinfo> {
         NI_MAXSERV,
         NI_NUMERICSERV);
     if (r != 0) {
-      return fmt::format_to(ctx.out(), "?UNKNOWN?");
+      // if we can't resolve the hostname, display the IP address
+      if (addr.ai_family == AF_INET) {
+        struct sockaddr_in* psai = (struct sockaddr_in*)addr.ai_addr;
+        char ip[INET_ADDRSTRLEN];
+        if (inet_ntop(addr.ai_family, &(psai->sin_addr), ip, INET_ADDRSTRLEN) !=
+            NULL) {
+          return fmt::format_to(ctx.out(), "{}:{}", ip, psai->sin_port);
+        }
+      } else if (addr.ai_family == AF_INET6) {
+        struct sockaddr_in6* psai = (struct sockaddr_in6*)addr.ai_addr;
+        char ip[INET6_ADDRSTRLEN];
+        if (inet_ntop(
+                addr.ai_family, &(psai->sin6_addr), ip, INET6_ADDRSTRLEN) !=
+            NULL) {
+          return fmt::format_to(ctx.out(), "[{}]:{}", ip, psai->sin6_port);
+        }
+      }
+      C10_THROW_ERROR(
+          DistNetworkError,
+          fmt::format(
+              "failed to format addr, unknown family={}", addr.ai_family));
     }
```

Comment:
We're seeing 
```
failed to format addr, unknown family=0
```
in a job which would otherwise run fine if we disable logging by TORCH_CPP_LOG_LEVEL=WARNING.

This is probably a bug - we shall not assume we're formatting a VALID addrinfo. When connection is being established there might be invalid addrinfo to be logged.

Reply:
We have not identified exactly what causes the invalid addrinfo to appear, but I'm sending https://github.com/pytorch/pytorch/pull/137745/files and I believe it is a good fix regardless. In general, formatter should avoid crashing as much as possible.